legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java

```diff
@@ -0,0 +1,725 @@
+ * <li>:remove the messaging subsystem</li>
+ * </ul>
+ * <p/>
+ * The companion <code>:describe-migration</code> operation will return a list of all the actual operations that would be
+ * performed during the invocation of the <code>:migrate</code> operation.
+ * <p/>
+ * Note that all new operation addresses are generated for standalone mode. If this is a domain mode server
+ * then the addresses are fixed after they have been generated
+ *
+ * @author <a href="http://jmesnil.net/">Jeff Mesnil</a> (c) 2015 Red Hat inc.
+ * @author Stuart Douglas
+ */
+
+public class WebMigrateOperation implements OperationStepHandler {
+
+    private static final String UNDERTOW_EXTENSION = "org.wildfly.extension.undertow";
+    private static final String IO_EXTENSION = "org.wildfly.extension.io";
+
+    private static final String REALM_NAME = "jbossweb-migration-security-realm";
+
+    private static final OperationStepHandler DESCRIBE_MIGRATION_INSTANCE = new WebMigrateOperation(true);
+    private static final OperationStepHandler MIGRATE_INSTANCE = new WebMigrateOperation(false);
+    public static final PathElement DEFAULT_SERVER_PATH = pathElement(Constants.SERVER, "default");
+
+    private final boolean describe;
+
+    private WebMigrateOperation(boolean describe) {
+
+        this.describe = describe;
+    }
+
+    static void registerOperations(ManagementResourceRegistration registry, ResourceDescriptionResolver resourceDescriptionResolver) {
+        registry.registerOperationHandler(new SimpleOperationDefinitionBuilder("migrate", resourceDescriptionResolver)
+                        .setRuntimeOnly()
+                        .setAccessConstraints(SensitiveTargetAccessConstraintDefinition.READ_WHOLE_CONFIG)
+                        .build(),
+                WebMigrateOperation.MIGRATE_INSTANCE);
+        registry.registerOperationHandler(new SimpleOperationDefinitionBuilder("describe-migration", resourceDescriptionResolver)
+                        .setReplyType(ModelType.LIST).setReplyValueType(ModelType.OBJECT)
+                        .setRuntimeOnly()
+                        .setAccessConstraints(SensitiveTargetAccessConstraintDefinition.READ_WHOLE_CONFIG)
+                        .build(),
+                WebMigrateOperation.DESCRIBE_MIGRATION_INSTANCE);
+    }
+
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        if (!describe && context.getRunningMode() != RunningMode.ADMIN_ONLY) {
+            throw UndertowLogger.ROOT_LOGGER.migrateOperationAllowedOnlyInAdminOnly();
+        }
+
+        // node containing the description (list of add operations) of the legacy subsystem
+        final ModelNode legacyModelAddOps = new ModelNode();
+        //we don't preserve order, instead we sort by address length
+        //TODO: is this ok in every case?
+        final Map<PathAddress, ModelNode> migrationOperations = new TreeMap<>(new Comparator<PathAddress>() {
```

Comment:
fwiw, for the messaging subsystem, I used a LinkedHashMap to rely on the order of the operations returned by the :describe operations.